workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/DialogCollator.kt

```diff
@@ -0,0 +1,214 @@
+   * to display [overlay], or `null` if there is none.
+   */
+  fun find(overlay: Overlay): DialogSession?
+}
+
+/**
+ * Provided by [LayeredDialogSessions] to [DialogCollator.scheduleUpdates].
+ * Knows how to create a [DialogSession] for [overlay], or to update an existing
+ * one.
+ *
+ * @param overlay rendering to be shown in a new or existing Dialog window
+ *
+ * @param doUpdate function to create or update a [DialogSession] to display
+ * [overlay]. Provided with an [OldSessions] instance that provides access to the
+ * existing session, if any, and a `covered: Boolean` param that indicates
+ * if there is a [ModalOverlay] above [overlay]
+ */
+@WorkflowUiExperimentalApi
+internal class DialogSessionUpdate(
+  val overlay: Overlay,
+  val doUpdate: (
+    oldSessions: OldSessions,
+    covered: Boolean
+  ) -> DialogSession
+)
+
+/**
+ * Init method called at the start of [LayeredDialogSessions.update].
+ * Ensures that there is a single [DialogCollator] instance shared by
+ * an entire recursive hierarchy of [LayeredDialogSessions] for each
+ * update pass.
+ *
+ * Each call to [establishDialogCollator] must be matched with a call to
+ * [DialogCollator.scheduleUpdates]. This balance lets us know when a set of recursive
+ * [LayeredDialogSessions.update] calls is complete, so that we can manage
+ * the entire set of [Dialog][android.app.Dialog] windows.
+ *
+ * Returns a new [ViewEnvironment] with a [DialogCollator]
+ * value if none was found, otherwise returns self. Either way,
+ * the given [existingSessions] set is added to the pool
+ * that will be processed by the new or existing [DialogCollator].
+ *
+ * @param id unique runtime-only id used to pair calls to [establishDialogCollator]
+ * and [DialogCollator.scheduleUpdates]
+ *
+ * @param existingSessions [DialogSession] instances that were running before the pending update
+ */
+@WorkflowUiExperimentalApi
+internal fun ViewEnvironment.establishDialogCollator(
+  id: UUID,
+  existingSessions: List<DialogSession>
+): ViewEnvironment {
+  val collatorOrNull = map[DialogCollator]
+  val collator = (collatorOrNull as? DialogCollator) ?: DialogCollator()
+
+  collator.expectedUpdates++
+  collator.establishedSessions.add(
+    SessionAndViewStateMap(id, existingSessions)
+  )
+
+  return if (collatorOrNull == null) this + (DialogCollator to collator) else this
+}
+
+/**
+ * Resource shared by a recursive set of [LayeredDialogSessions], used to
+ * ensure that the windows they define are stacked in the correct order and
+ * stay that way across updates.
+ *
+ * Android notoriously doesn't give us any control over the z order of windows
+ * other than taking care to [show][android.app.Dialog.show] them in the right
+ * order, so keeping a pile of windows in sync with the shifting order of the
+ * defining list of [Overlay] rendering models is hard. [DialogCollator] manages
+ * that process. On each update it builds complete knowledge of the existing stack
```

Comment:
Looks like I didn't finish this paragraph. :typing:

Reply:
```
 * Singleton resource shared by a recursive set of [LayeredDialogSessions], used to
 * ensure that the windows they define are stacked in the correct order and stay that
 * way across updates.
 *
 * Android notoriously doesn't give us any control over the z order of windows
 * other than taking care to [show][android.app.Dialog.show] them in the right
 * order, so keeping a pile of windows in sync with the shifting order of the
 * defining list of [Overlay] rendering models is hard. [DialogCollator] manages
 * that process.
 *
 * When a workflow UI tree is updated (that is, each time
 * [WorkflowLayout][com.squareup.workflow1.ui.WorkflowLayout.show] is called), a shared
 * [DialogCollator] is used across nested [LayeredDialogSessions] to coordinate their work.
 * Specifically, a [DialogCollator] instance is put in place in the [ViewEnvironment]
 * when [LayeredDialogSessions.update] calls [ViewEnvironment.establishDialogCollator].
 * The [LayeredDialogSessions] loads its [DialogCollator] with functions to create or update
 * managed [Dialog][android.app.Dialog] instances.
 *
 * Any recursive calls to [LayeredDialogSessions.update] receive the same [DialogCollator]
 * when they make their own calls to [ViewEnvironment.establishDialogCollator], and so
 * enqueue their updates in the same place.
 *
 * When control returns to the first [LayeredDialogSessions.update] stack frame, all of the
 * updates that were enqueued with the shared [DialogCollator] are executed in a single
 * pass. Because this [DialogCollator] has complete knowledge of the existing stack
 * of `Dialog` windows and all updates, it is able to decide if any existing instances need to be
 * [dismissed][android.app.Dialog.dismiss] and [re-shown][android.app.Dialog.show]
 * to keep them in the correct order.
```